### DISCUSSION TITLE
Invalidate cache based on entity not the query key.

### DISCUSSION THREAD

**Discussion by esuca** [archimedes-projects/archimedes-js#34] [Ideas]
A long time ago, I learned how to use [RTK Query](https://redux-toolkit.js.org/rtk-query/overview):

> RTK Query is a powerful data fetching and caching tool. It is designed to simplify common cases for loading data in a web application, eliminating the need to hand-write data fetching & caching logic yourself.
> 
> RTK Query is an optional addon included in the Redux Toolkit package, and its functionality is built on top of the other APIs in Redux Toolkit.

RTK Query also uses the query/command pattern to work with the cache.
I was very interested in their concept of "cache tags" to invalidate data. 

**Each query provide an array of tags, for example: ['Post', 'User'].**
**And the commands**, in the case of RTK Query, mutations, you **specify which tags to invalidate.**

> RTK Query uses the concept of 'tags' to determine whether a mutation for one endpoint intends to invalidate some data that was provided by a query from another endpoint.
> 
> If cache data is being invalidated, it will either refetch the providing query (if components are still using that data) or remove the data from the cache.

The 'Tag' naming is a bit strange, I think more about it as if it were an **Entity/Model/Resource**. 

```ts
@QueryEntities([ "User" ])
class GetUsersQry extends Qry<User[]> {

  execute() {
     //  ....
  }
}
```

```ts
@InvalidateQueryEntities([ "User" ])
class AddUserCmd extends Command<User> {

  execute() {
     //  ....
  }
}
```

So, the CacheInvalidations object needs to change, maybe something like this:

```ts
// from:
const currentCacheInvalidationsObj = {
 "AddUserCmd": ["GetUsersQry"]
}

// to:
const newCacheInvalidationsObj = {
  "AddUserCmd": ["User"]
}
```

The example is very basic and easy to understand. I have opened this discussion because in larger projects, it becomes a bit confusing to know which command invalidates which query. 

A common mistake is to add a new query and forget to update the invalidation list of all the commands that have to do with that query. The RTK Query solution I believe reduces the frequency of that mistake, and at the same time makes it easier to identify what we invalidate.

Also, RTK Query solution, answers more to the question **"What does the query contain?"**, than to the question **"Which query do we want to invalidate?"** which is the question we ask ourselves with the current implementation.

¿What do you think?

**Comment by ArturoRodriguezRomero** [comment]
Makes sense to me. A bit more open and flexible than the current solution.
Maybe it'd be useful to type the "tags".

`const newCacheInvalidationsObj = {
  "AddUserCmd": [QueryResource.USER]
}`

**Comment by z0r3f** [comment]
Great Concept.
I think it is better to list the updaters (Cmd) with entities affect than to record the victims (Qry). Agree.

One add-on. Only specify which tags to invalidate as you have mentioned. Not all the resources in the domain.
In your example, if `User` has `Address` and it has `Country` (no updatable), we only will to invalid `User` and `Address`; no `Country`.
